[bhavinaa] IP

bhavinaa opened the pull request:
# 🚀 Bhavs  

> “Lost time is never found again” – Benjamin Franklin.  

## 📌 What is Bhavs?  
Bhavs is a **task management and productivity tool** designed to help users stay organized, meet deadlines, and efficiently track their goals. Whether you're a student, professional, or entrepreneur, Bhavs provides an intuitive interface to manage tasks seamlessly.  

## Overview  
Bhavs is a lightweight, **text-based** productivity application 🚀 that allows users to:  
- **Add and manage tasks** effortlessly  
- **Track deadlines** to never miss important dates  
- **Set reminders** for upcoming tasks  
- **Filter tasks** based on priority, due date, or completion status  

## Motivation  
The idea behind Bhavs was born out of the **need for a simple yet effective** task management solution. Many existing tools are either too complex or lack the *flexibility* needed for different workflows.  

This is a ~~simple~~ very simple markdown tool.  

Our key motivations:  
1. **Simplicity & Usability** – Creating an easy-to-use tool without unnecessary complexity  
2. **Productivity & Focus** – Helping users organize their tasks without distractions  
3. **Efficiency & Speed** – A lightweight and fast application for quick task management  

## 🛠 Features  
- ✅ Task management  
- ✅ Deadline tracking  
- ✅ Reminders  
- ✅ Task filtering  

## Sample Code  
```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello, world!");
    }
}
```

## Author

[More about the Author](https://github.com/bhavinaa)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

bhavinaa committed:
Level 0. Rename, Greet, Exit

bhavinaa committed:
Level 1. Echo

bhavinaa committed:
Level 2. Add, List

bhavinaa committed:
added the task class

bhavinaa committed:
started on level 3, with an option to toggle

bhavinaa committed:
final level 3 work

bhavinaa committed:
refactored the code and broke it down to functions

bhavinaa committed:
rough qns 4

bhavinaa committed:
almost done with level 4

bhavinaa committed:
added the level 4, and level 6, as i try to see if i can add more error/exception checks

bhavinaa committed:
prev version had an error.. it does not compile

bhavinaa committed:
fixed the mark list as i was double subtracting it

bhavinaa committed:
made the default test cases

bhavinaa committed:
edited grammer

bhavinaa committed:
branch Level-7

bhavinaa committed:
Merge branch-Level-7 with Level 7 implementation

bhavinaa committed:
add the command feature to show all possible commands

bhavinaa committed:
added the UI class to increase OOP

bhavinaa committed:
added the storage class

bhavinaa committed:
added the parser in the code

bhavinaa committed:
added more OOP

bhavinaa committed:
Merged A-MoreOOP refactoring into main

bhavinaa committed:
Merge branch 'A-MoreOOP'

bhavinaa committed:
draft of the time class

bhavinaa committed:
to merge later

bhavinaa committed:
refactored to packages

bhavinaa committed:
Merge remote-tracking branch 'origin/add-gradle-support'

bhavinaa committed:
add gradle support

bhavinaa committed:
test work in progress

bhavinaa committed:
added the Junit tests

bhavinaa committed:
add the events file

bhavinaa committed:
add JavaDocs

bhavinaa committed:
wrote javaDocs

bhavinaa committed:
Add JavaDocs for the events class

bhavinaa committed:
Merge branch 'branch-A-JavaDoc'

bhavinaa committed:
added JavaDocs

bhavinaa committed:
Resolved merge conflict in Events.java

bhavinaa committed:
added the refactored code

bhavinaa committed:
refactored code

bhavinaa committed:
Merge branch 'branch-A-CodingStandard'

bhavinaa committed:
refactored the classes

bhavinaa committed:
add all the relevant stuff

bhavinaa committed:
added Level-9

bhavinaa committed:
Merging branch-Level-9 into master

bhavinaa committed:
Check coding style

bhavinaa committed:
Merge branch 'git-standard'

lesterlimjj commented on src/main/java/bhavs/tasks/Deadlines.java:
```diff
@@ -0,0 +1,29 @@
+
+import bhavs.utils.Time;
+
+public class Deadlines extends Task {
```
Should we name the class as Deadlines when we are creating a single Deadline?

lesterlimjj commented on src/main/java/bhavs/tasks/Events.java:
```diff
@@ -0,0 +1,40 @@
+ * the way it is read from the file
+ **/
+
+public class Events extends Task {
```
Same comment as the one given to deadline

lesterlimjj commented on src/main/java/bhavs/tasks/ToDos.java:
```diff
@@ -0,0 +1,22 @@
+package bhavs.tasks;
+
+public class ToDos extends Task {
```
Same comment as the one given to deadline

lesterlimjj commented on src/main/java/bhavs/tasks/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+public class TaskList {
+
+    private List<Task> taskList = new ArrayList<>();
+    public Storage storedIn;
```
Class variables should not be declared as public. Perhaps we can have storage be a field in main and perform its function calls there instead?

lesterlimjj replied:
Also the name storedIn implies that this variable tells where the data is store (e.g. a file path) which can be confusing. Giving it a name such as store or fileStore would be more appropriate.

lesterlimjj commented on src/main/java/bhavs/utils/Time.java:
```diff
@@ -0,0 +1,67 @@
+import java.time.format.DateTimeParseException;
+import java.util.Scanner;
+
+public class Time {
```
I like the idea of a Time class!

jeronkk replied:
Nice use of the Time class to add more OOP to your application

lesterlimjj commented on src/main/java/bhavs/utils/UI.java:
```diff
@@ -0,0 +1,155 @@
+the UI class contains all the helper functions that allows you to interact with
+the system and access items from the storage
+*/
+public class UI {
```
Your UI class is handling more responsibilities than just displaying UI elements. Consider extracting out these extra responsibilities to other classes. 

A suggestion would be to have your storage, scanner, parser and taskList fields be in main instead and perform their work there.

lesterlimjj commented on src/main/java/bhavs/utils/UI.java:
```diff
@@ -0,0 +1,155 @@
+        return scanner.nextLine();
+    }
+
+    public void processComands() {
```
Perhaps this could be in Parser instead, since it already handles commands related to task creation?

jeronkk replied:
This is a typo, it should be processCommands().

lesterlimjj commented on src/main/java/bhavs/utils/UI.java:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    public Task make_correct_entry(String userCommand) {
```
Your function names should be in camelCase.

lesterlimjj commented on src/main/java/bhavs/tasks/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return The valid task index.
+     */
+    public int getTaskIndex(Scanner scanner, String prompt, int listSize) {
```
It's a cool design choice to have your user be prompted for index until a correct one is given when CRUD operations(e.g. mark, delete). However, if a user were to input it by mistake, they have no way of cancelling the command except for inputting a valid index.

lesterlimjj commented on src/main/java/bhavs/utils/Storage.java:
```diff
@@ -0,0 +1,80 @@
+public class Storage {
+
+    private String filePath;
+    public TaskList taskList;
```
Class fields should not be public. Same suggestion of moving these field to main and handling the work there is given.

lesterlimjj commented on src/main/java/bhavs/utils/Time.java:
```diff
@@ -0,0 +1,67 @@
+    private static final DateTimeFormatter OUTPUT_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy, hh:mm a");
+    private LocalDateTime dateTime;
+
+    public Time(String input, boolean isInteractive) {
```
Extending the function of a constructor to check for prompt or parsing is interesting but not really the intended use of a constructor. Perhaps this logic can be handled elsewhere?

lesterlimjj reviewed (commented):
I would like to mention that this is my 2nd PR review so I am not putting much focus on the coding standard, but instead code quality (with a focus on naming). The code here has some interesting and cool design choice (e,g, prompting user for a correct input after a command is locked in) but needs to be careful of over extending the functionality classes.

jeronkk commented on src/main/java/bhavs/tasks/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+            return;
+        }
+
+        int taskIndex = getTaskIndex(scanner, "Which task number to unmark?", taskList.size());
```
Both markTask and unmarkTask have similar logic, maybe implementing a helper method to reduce the number of lines?

jeronkk commented on src/main/java/bhavs/utils/Parser.java:
```diff
@@ -0,0 +1,44 @@
+
+import bhavs.tasks.Deadlines;
+import bhavs.tasks.Events;
+import bhavs.tasks.*;
```
Best is to not use Wildcard (*) imports as it can lead to ambiguity and unnecessary imports.

jeronkk commented on src/main/java/bhavs/tasks/Task.java:
```diff
@@ -0,0 +1,74 @@
+
+    public boolean isCompleted() {
+        return isDone;
+    }
```
There are some missing javadoc comments here.

jeronkk reviewed (commented):
LGTM. Just some nits to fix.

bhavinaa committed:
Update README.md

bhavinaa committed:
Update README.md

bhavinaa committed:
Update README.md for version 1.0

bhavinaa committed:
Add and Run the checkstyle

bhavinaa committed:
Merge branch 'checkStyle'

bhavinaa committed:
Add the launcher class and start GUI migration

bhavinaa committed:
Add the FXML file - test

bhavinaa committed:
works more now

bhavinaa committed:
Refactored to remove unnecessary code and made it state persistent

bhavinaa committed:
Resolve Merge Conflicts

bhavinaa committed:
Save changes before switching branches

bhavinaa committed:
Update sample data

bhavinaa committed:
Resolved merge conflicts

bhavinaa committed:
Update the gradle

bhavinaa committed:
Add Assertions

bhavinaa committed:
Refactor the UI class

bhavinaa committed:
Merge the Refactored Code

bhavinaa committed:
Refactor the Code

bhavinaa committed:
Merge source-branch into target-branch

bhavinaa committed:
Add the processCommand function

bhavinaa committed:
Merge source-branch into Master

bhavinaa committed:
Add Assertions

bhavinaa committed:
Refactor the code

bhavinaa committed:
Merge pull request #2 from bhavinaa/branch-A-codeQuality

Refactor the code

bhavinaa committed:
Change the data

bhavinaa committed:
Merge branch 'master' into branch-A-Assertions

bhavinaa committed:
Merge pull request #3 from bhavinaa/branch-A-Assertions

Add Assertions

bhavinaa committed:
Added GitHub Actions CI workflow

bhavinaa committed:
Merge branch 'branch-A-Assertions' of github.com:bhavinaa/ip into branch-A-Assertions

bhavinaa committed:
Merge pull request #4 from bhavinaa/branch-A-Assertions

Add the CI/CD

bhavinaa committed:
Added GitHub Actions CI workflow

bhavinaa committed:
Update the .yml file

bhavinaa committed:
Process Commands

bhavinaa committed:
Update the .yml file

bhavinaa committed:
Update the .yml file

bhavinaa committed:
Update the .yml file

bhavinaa committed:
Update the .yml file

bhavinaa committed:
Refactor TaskManager class

Refactored the UI class to separatecommand processing logic.
Extracted CommandProcessor into its own class
Improved method naming for better readability
Updated relevant tests to reflect the changes

This changes makes the code more modular and easier to maintain.

bhavinaa committed:
Add the Extension Feature C-Sort

bhavinaa committed:
Change the .yml file

bhavinaa committed:
Disable checkstyle

bhavinaa committed:
Disable Checkstyle

bhavinaa committed:
Remove CheckStyle

bhavinaa committed:
Add the updated the GUI

bhavinaa committed:
Refactor code for submission

bhavinaa committed:
Add Screenshot of UI

bhavinaa committed:
Update readme

bhavinaa committed:
Update readme

bhavinaa committed:
Update readme

bhavinaa committed:
Update readme

bhavinaa committed:
Refactor readme

bhavinaa committed:
Fix compatability with Linux

bhavinaa committed:
Update README.md

bhavinaa committed:
Update Bhavs.java

removed unnecessary code
